src/instrumentationESM.ts

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License.
+
+// Import Azure Functions types
+import type * as AzFunc from '@azure/functions';
+// Import OpenTelemetry APIs for context propagation
+import { context as otelContext, propagation } from '@opentelemetry/api';
+// Import OpenTelemetry severity levels for logging
+import { SeverityNumber } from '@opentelemetry/api-logs';
+// Import OpenTelemetry instrumentation base classes
+import {
+    InstrumentationBase,
+    InstrumentationConfig,
+    InstrumentationNodeModuleDefinition,
+} from '@opentelemetry/instrumentation';
+// Import instrumentation version constant
+import { version } from './constants';
+
+/**
+ * Azure Functions Instrumentation class specifically designed for ESM scenarios.
+ * Extends OpenTelemetry's InstrumentationBase to integrate with Azure Functions.
+ */
+export class AzureFunctionsInstrumentationESM extends InstrumentationBase {
+    // Array to store disposables for cleanup during unpatching
+    private azFuncDisposable: AzFunc.Disposable[] = [];
+
+    /**
+     * Constructor initializes the instrumentation with a given configuration.
+     * @param config Optional instrumentation configuration.
+     */
+    constructor(config: InstrumentationConfig = {}) {
+        super('@azure/functions-opentelemetry-instrumentation', version, config);
+    }
+
+    /**
+     * Initializes the instrumentation by defining how to patch and unpatch the Azure Functions module.
+     * This method is called automatically by OpenTelemetry instrumentation lifecycle.
+     */
+    protected init() {
+        const instrumentationNodeModuleDefinition = new InstrumentationNodeModuleDefinition(
+            '@azure/functions', // Target module to instrument
+            ['^4.5.0'], // Supported versions
```

Comment:
If we release a new major library version, this will not be picked up.
I suggest using ['>=4.5.0']

Reply:
This meant to be temporary until @opentelemetry group fix it

Reply:
https://github.com/Azure/azure-functions-pyfx-planning/issues/776